### PR TITLE
For Non Apple platforms, allow a blank master key for wrapping.

### PR DESCRIPTION
**Pull request by nmahendru**
#### 0a79bb9adb4cbe8144ea6e3b133b6bcf6ca7ccc1
<pre>
For Non Apple platforms, allow a blank master key for wrapping.
<a href="https://bugs.webkit.org/show_bug.cgi?id=275890">https://bugs.webkit.org/show_bug.cgi?id=275890</a>
<a href="https://rdar.apple.com/136807212">rdar://136807212</a>

Reviewed by Carlos Garcia Campos.

In <a href="https://commits.webkit.org/276197@main">https://commits.webkit.org/276197@main</a>, the flow for Wrapping/Unwrapping of
CryptoKeys for local storage was changed to early exit if the webCryptoMasterKey is NULL.

For non apple ports the wrap/unwrap operations are passthrough and don&apos;t actually use the master key.
Now, because the early-exit behavior, for non-apple ports, the wrap/unwrap operations are now returning null in cases where master key is null/zero length buffer.
Earlier they were not. That has been investigated as the likely cause for the Whatsapp Spinner in this bug.

This change brings back that behavior for non apple ports.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::wrapCryptoKey):
(WebKit::WebPageProxy::unwrapCryptoKey):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::wrapCryptoKey):
(WebKit::WebProcessProxy::unwrapCryptoKey):

Canonical link: <a href="https://commits.webkit.org/284377@main">https://commits.webkit.org/284377@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/288830fa7e0dff53cb493f679972ef645ab0ff39

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/69119 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48519 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21791 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/73200 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/20277 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/56320 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/20126 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/55007 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13454 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72185 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44286 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59669 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35484 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40954 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18651 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62897 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17444 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74911 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/13101 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16681 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/62659 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/13139 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59752 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62558 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15364 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10561 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4171 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/44323 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45396 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46592 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/45138 "Built successfully") | | |

### CODE REVIEW

**Review by carlosgcampos — APPROVED** [review]

**Comment by mcatanzaro** [comment]
Thank you!

(I wonder what the master key is....)

File: Source/WebKit/UIProcess/WebPageProxy.cpp [hunk 1]
```diff
@@ -12464,22 +12464,28 @@ void WebPageProxy::getWebCryptoMasterKey(CompletionHandler<void(std::optional<Ve
...
-            if (wrapSerializedCryptoKey(*masterKey, key, wrappedKey))
-                return completionHandler(WTFMove(wrappedKey));
-        }
+#if PLATFORM(COCOA)
+        if (!masterKey)
+            return completionHandler(std::nullopt);
+#endif
+        Vector<uint8_t> wrappedKey;
+        const Vector<uint8_t> blankMasterKey;
+        if (wrapSerializedCryptoKey(masterKey.value_or(blankMasterKey), key, wrappedKey))
```
Comment (achristensen07) [inline Source/WebKit/UIProcess/WebPageProxy.cpp:12473]: Can't you just do masterKey.value_or({ }) ?

Comment (nmahendru) [reply Source/WebKit/UIProcess/WebPageProxy.cpp:12473]: Giving it a name that specifies we are passing a blank masterKey, for readability, was my idea.

**Comment by nmahendru** [comment]
> Thank you!
> 
> (I wonder what the master key is....)

There isn't one for Non Apple platforms. https://github.com/WebKit/WebKit/blob/62cb2d15cfc80d1029a7fee940fcfa07d89bf0e8/Source/WebCore/crypto/gcrypt/SerializedCryptoKeyWrapGCrypt.cpp#L47 <-- is not using anything that is passed in. It' basically setup so that encrypt/decrypt is a NOP

**Comment by mcatanzaro** [comment]
What is the master key used for? This isn't part of the WebCrypto spec, and I can't find any explanations with a web search. So it's hard to know whether it's a problem for it to be empty.

**Comment by nmahendru** [comment]
> What is the master key used for? This isn't part of the WebCrypto spec, and I can't find any explanations with a web search. So it's hard to know whether it's a problem for it to be empty.

It's not part of the WebStandard. It's a quirk in WebKit. 
Some context:
https://github.com/WebKit/WebKit/blob/b67dcdc4fe7d594a08e1cd1d8160f525395d2f38/Source/WebCore/crypto/SerializedCryptoKeyWrap.h#L35
https://bugs.webkit.org/show_bug.cgi?id=128680

**Comment by szewai** [comment]
> What is the master key used for? This isn't part of the WebCrypto spec, and I can't find any explanations with a web search. So it's hard to know whether it's a problem for it to be empty.

I think it's used for wrapping the keys for storage, see Source/WebCore/crypto/SerializedCryptoKeyWrap.h.
It looks like the impl of these functions are no-op 

> > What is the master key used for? This isn't part of the WebCrypto spec, and I can't find any explanations with a web search. So it's hard to know whether it's a problem for it to be empty.
> 
> It's not part of the WebStandard. It's a quirk in WebKit. Some context:
> 
> https://github.com/WebKit/WebKit/blob/b67dcdc4fe7d594a08e1cd1d8160f525395d2f38/Source/WebCore/crypto/SerializedCryptoKeyWrap.h#L35
> 
> 
> https://bugs.webkit.org/show_bug.cgi?id=128680

If the other ports don't need the wrap / unwrap, it seems they shouldn't be doing the messaging between UI process and Web Process at all (i.e. `WebPageProxy::wrapCryptoKey` should be a Cocoa platform specific function, instead of making `WebPageProxy::wrapCryptoKey` platform specific).

**Comment by szewai** [comment]
> > > What is the master key used for? This isn't part of the WebCrypto spec, and I can't find any explanations with a web search. So it's hard to know whether it's a problem for it to be empty.
> > 
> > 
> > It's not part of the WebStandard. It's a quirk in WebKit. Some context:
> > https://github.com/WebKit/WebKit/blob/b67dcdc4fe7d594a08e1cd1d8160f525395d2f38/Source/WebCore/crypto/SerializedCryptoKeyWrap.h#L35
> > 
> > https://bugs.webkit.org/show_bug.cgi?id=128680
> 
> If the other ports don't need the wrap / unwrap, it seems they shouldn't be doing the messaging between UI process and Web Process at all (i.e. `WebPageProxy::wrapCryptoKey` should be a function only exists on Cocoa platforms, instead of making `WebPageProxy::wrapCryptoKey` platform specific).

**Comment by mcatanzaro** [comment]
> It's not part of the WebStandard. It's a quirk in WebKit. Some context:

Looks like it's used to encrypt a secret in the system keyring. This would be redundant on Linux, because the keyring is already encrypted.

> If the other ports don't need the wrap / unwrap, it seems they shouldn't be doing the messaging between UI process and Web Process at all (i.e. WebPageProxy::wrapCryptoKey should be a function only exists on Cocoa platforms, instead of making WebPageProxy::wrapCryptoKey platform specific).

Probably....

**Comment by aperezdc** [comment]
Backported into the 2.46 branch as commit c369b75afa18e9388a19eb422b7b3f98bb8d56c3

**Pull request merged**